src/instance.rs

```diff
@@ -38,6 +40,190 @@ pub unsafe trait PyNativeType: Sized {
+    fn deref(&self) -> &Py2<'py, PyAny> {
+        ::std::convert::AsRef::as_ref(self)
+    }
+}
+
+impl<'py, T> AsRef<Py2<'py, PyAny>> for Py2<'py, T> {
+    fn as_ref(&self) -> &Py2<'py, PyAny> {
+        // Safety: &Py2<T> has the same layout as &Py2<PyAny>
+        unsafe { std::mem::transmute(self) }
+    }
+}
+
+impl<T> Clone for Py2<'_, T> {
+    fn clone(&self) -> Self {
+        Self(self.0, ManuallyDrop::new(self.1.clone_ref(self.0)))
+    }
+}
+
+impl<T> Drop for Py2<'_, T> {
+    fn drop(&mut self) {
+        unsafe { ffi::Py_DECREF(self.1.as_ptr()) }
+    }
+}
+
+impl<'py, T> Py2<'py, T> {
+    /// Returns the GIL token associated with this object.
+    pub fn py(&self) -> Python<'py> {
+        self.0
+    }
+
+    /// Returns the raw FFI pointer represented by self.
+    ///
+    /// # Safety
+    ///
+    /// Callers are responsible for ensuring that the pointer does not outlive self.
+    ///
+    /// The reference is borrowed; callers should not decrease the reference count
+    /// when they are finished with the pointer.
+    #[inline]
+    pub fn as_ptr(&self) -> *mut ffi::PyObject {
+        self.1.as_ptr()
+    }
+
+    /// Returns an owned raw FFI pointer represented by self.
+    ///
+    /// # Safety
+    ///
+    /// The reference is owned; when finished the caller should either transfer ownership
+    /// of the pointer or decrease the reference count (e.g. with [`pyo3::ffi::Py_DecRef`](crate::ffi::Py_DecRef)).
+    #[inline]
+    pub fn into_ptr(self) -> *mut ffi::PyObject {
+        let ptr = (self.1).0.as_ptr();
+        std::mem::forget(self);
+        ptr
+    }
+
+    /// Internal helper to convert e.g. &'a &'py PyDict to &'a Py2<'py, PyDict> for
+    /// backwards-compatibility during migration to removal of pool.
+    #[doc(hidden)] // public and doc(hidden) to use in examples and tests for now
+    pub fn borrowed_from_gil_ref<'a>(gil_ref: &'a &'py T::AsRefTarget) -> &'a Self
+    where
+        T: PyTypeInfo,
+    {
+        // Safety: &'py T::AsRefTarget is expected to be a Python pointer,
+        // so &'a &'py T::AsRefTarget has the same layout as &'a Py2<'py, T>
+        unsafe { std::mem::transmute(gil_ref) }
+    }
+
+    /// Internal helper to get to pool references for backwards compatibility
+    #[doc(hidden)] // public and doc(hidden) to use in examples and tests for now
+    pub fn as_gil_ref(&'py self) -> &'py T::AsRefTarget
+    where
+        T: PyTypeInfo,
+    {
+        unsafe { self.py().from_borrowed_ptr(self.as_ptr()) }
+    }
+
+    /// Internal helper to get to pool references for backwards compatibility
+    #[doc(hidden)] // public but hidden, to use for tests for now
+    pub fn into_gil_ref(self) -> &'py T::AsRefTarget
+    where
+        T: PyTypeInfo,
+    {
+        unsafe { self.py().from_owned_ptr(self.into_ptr()) }
+    }
+
+    pub(crate) fn into_non_null(self) -> NonNull<ffi::PyObject> {
+        let ptr = (self.1).0;
+        std::mem::forget(self);
+        ptr
```

Comment:
My personal preference would be to just use 
```suggestion
        ManuallyDrop(self).1.0
```

Reply:
👍 will change to that, it's a lot tidier and hopefully both optimize away.